### PR TITLE
Implement Document create (POST) endpoint

### LINKED ISSUE #19: ODRC: implement Document creation/upload
**Acceptance criteria**

Copied/moved from #2

- Het format wordt gevalideerd wanneer het bestand geüpload wordt.
- Via een API kan een document (data-object) gecreëerd, geraadpleegd, gemuteerd en verwijderd worden.
- Bij een document moet exact één bestand (bijvoorbeeld een PDF) worden opgeslagen.
- De creatie, mutatie of verwijdering van een document wordt gelogd (zie #16)
- De API specificatie is bijgewerkt (ReDoc, Swagger)
- De documentatie is bijgewerkt (Read the Docs)

## Document creation + uploads

We make use of the file parts mechanism of the Documenten API, always.

* ODRC will proxy to the underlying documenten API

### Preparing a document upload

The client must pass the necessary metadata and then based on the response of that,
cut up the file upload in parts that can be submitted individually.

The request body schema of a Document POST would look something like:

```yaml
Document:
  type: object
  required:
    - identifier
  properties:
    publicatie:  # UUID of the publication it belongs to
      type: string
      format: uuid
    identifier:  # the 'primary' identifier
      type: string
    creatiedatum:
      type: string
      format: date
    officieleTitel:  # DiWoo doesn't seem to apply a max length
      type: string
    verkorteTitel:
      type: string
    omschrijving:
      type: string
    # from waardelijst -> expose options in separate endpoint (!)
    # POST (write) operations should be able to just provide the identifier IRI instead of this complex object if ICATT desires this
    bestandsformaat:  
      type: object
      properties:
        identifier:  # IRI from waardelijst
          type: string
          format: uri
        mimeType:  # e.g. application/pdf
          type: string
        naam:
          type: string  # e.g. "PDF"
```

This translates to a request of ODRC -> Documenten API with schema:

```yaml
EnkelvoudigInformatieobject:
  type: object
  properties:
    identificatie:  # primary identifier of Document OR let the Documenten API generate one?
      type: string
    bronorganisatie:  # fixed, global configuration parameter in ODRC initially, could become 'smart' in the future
      type: string
    creatiedatum:  # taken from Document.creatiedatum
      type: string
      format: date
    titel:
      type: string
      minLength: 1
      maxLength: 200
    auteur:  # taken from Document.publicatie.organization
      type: string
    status:
      const: definitief  # archiving will move this to gearchiveerd, later
    formaat:  # derived from Document.bestandsformaat
      type: string
    taal:  # derived from Document.taal -> convert to/from ISO 639-2/B
      type: string
      enum:
        - dut
        - eng
    bestandsnaam:  # taken from Document.bestandsnaam
      type: string
    bestandsomvang:  # taken from Document.bestandsomvang, prepares the file parts
      type: number
    indicatieGebruiksrecht:
      const: false
    informatieobjecttype:  # points to /catalogi/api/v1/informatieobjecttypen/:uuid for the Document.publicatie.informatiecategorie
      type: string
      format: uri
```

The Documenten API will return a `lock` and list of BestandsDelen for upload, each
bestandsdeel will have the shape:

```yaml
BestandsDeel:
  type: object
  properties:
    url:  # URL to PUT to
      type: string
      format: uri
    volgnummer:
      type: integer
    omvang:
      type: integer
    voltooid:
      const: false
    lock: # ??
      type: string
```

The ODRC will then expose endpoints for these part uploads so the publication component
can upload the parts:

URL: `/api/v1/documenten/:uuid/bestandsdeel/:index`

A bestandsdeel wil simply be multipart/form-data, with the API key as auth header.

The request will be transformed by the ODRC, which adds the lock ID & JWT for the 
Documenten API, and streams the file part down to the Documenten API.

Once all parts are received, we unlock the created document.

**Tasks**

- [x] Implement API endpoint/serializer for `POST /api/v1/documenten`
    - Part of the metadata is stored in our database
    - Other metadata that we can store in Documenten API, we store there
    - Record the requested bestandsdelen from the Documenten API
    - Record the lock ID to finalize the upload/creation
    - Resolve the internal "Catalogi API" endpoint for the informatieobjecttype
- [x] Implement the API endpoint/serializer for `PUT /api/v1/documenten/:uuid/bestandsdelen/:index`
    - Use the stored lock ID & other configuration/metadata to all the Documenten API `/api/v1/bestandsdelen/:uuid`
    - If all parts are completely uploaded, unlock the document
    - Emit in the response body `documentFinalized: true|false` so that the client can be informed that they can refresh the document resource if needed, as file uploads will likely happen in parallel
    - Nice to have: check if we can make use of proxying/streaming (timebox: 1 day)
- ~~The first part of the document contains the magic bytes that can be used to validate against the `format` in the metadata. Check the upload validation in Open Forms for inspiration (and edge cases).~~ postponed to #36

### PR DESCRIPTION
**Pull request by sergei-maertens**
Fixes #19 (partially)

**Changes**

* Added global configuration
    - [x] Added model + admin
    - [x] Fix audit log mixin breaking initial config record create
    - [x] Add audit/admin smoketests
* [x] implement `POST /api/v1/documenten` endpoint
    * [x] Global config is used to create the document metadata, using `woo_publications.contrib.documenten_api.client`

### CODE REVIEW

File: src/woo_publications/config/models.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        "zgw_consumers.Service",
+        on_delete=models.PROTECT,
+        limit_choices_to={"api_type": APITypes.drc},
+        verbose_name=_("Documents API service"),
+        help_text=_(
+            "The service to use for new document uploads - the metadata and binary "
+            "content will be sent to this API."
+        ),
+        null=True,
+        blank=False,
```
Comment (bart-maykin) [inline src/woo_publications/config/models.py:30]: What is the reason why these two are different?

File: src/woo_publications/config/validators.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    Validates that a string value is a valid RSIN number by applying the
+    '11-proef' checking.
+
+    :param value: String object representing a presumably good RSIN number.
+    """
+    # Initial sanity checks.
+    validate_digits(value)
+    if len(value) != RSIN_LENGTH:
+        raise ValidationError(
+            "RSIN moet %s tekens lang zijn." % RSIN_LENGTH, code="invalid-length"
```
Comment (bart-maykin) [inline src/woo_publications/config/validators.py:27]: Any reason why you added multiple dutch error messages which aren't translatable?

Comment (sergei-maertens) [reply src/woo_publications/config/validators.py:27]: blind copy-paste from vng-api-common! I'm hopeful that when OAF/commonground-api-common is patched with 3.12 compatibility we can then just import the existing validator

File: src/woo_publications/config/models.py (hunk 1 above)
Comment (sergei-maertens) [reply src/woo_publications/config/models.py:30]: it's a required field, but django-solo creates the record if it doesn't exist yet, and making it not-nullable crashes with integrity error as there may not be any services defined yet while this record is being created.

`blank=False` ensures that form validation in the admin prevents you from leaving it empty

this pattern is one of the reasons I'm not too fond of django-solo anymore

File: src/woo_publications/config/validators.py (hunk 2 above)
Comment (bart-maykin) [reply src/woo_publications/config/validators.py:27]: It's probably smart to add a TODO comment then so we won't forget about it.

**Review by bart-maykin — APPROVED** [review]

**Pull request merged**